cdap-watchdog/src/main/java/co/cask/cdap/logging/appender/tms/TMSLogAppender.java

```diff
@@ -0,0 +1,267 @@
+    setName(APPENDER_NAME);
+    int queueSize = cConf.getInt(Constants.Logging.TMS_APPENDER_QUEUE_SIZE);
+    this.messageQueue = new ArrayBlockingQueue<>(queueSize);
+    this.tmsLogPublisher = new TMSLogPublisher(cConf, messageQueue, this, messagingService, queueSize);
+  }
+
+  @Override
+  public void start() {
+    tmsLogPublisher.startAndWait();
+    addInfo("Successfully started " + APPENDER_NAME);
+    super.start();
+  }
+
+  @Override
+  public void stop() {
+    tmsLogPublisher.stopAndWait();
+    addInfo("Successfully stopped " + APPENDER_NAME);
+    super.stop();
+  }
+
+  @Override
+  protected void appendEvent(LogMessage logMessage) {
+    logMessage.prepareForDeferredProcessing();
+    logMessage.getCallerData();
+
+    try {
+      messageQueue.put(logMessage);
+    } catch (InterruptedException e) {
+      addInfo("Interrupted when adding log message to queue: " + logMessage.getFormattedMessage());
+    }
+  }
+
+  // Based off of StringPartitioner, but that class can not be used in Standalone, as kafka dependencies do not exist
+  // in Standalone
+  @VisibleForTesting
+  static int partition(Object key, int numPartitions) {
+    return Math.abs(Hashing.md5().hashString(key.toString()).asInt()) % numPartitions;
+  }
+
+  /**
+   * Publisher service to publish logs to TMS asynchronously.
+   */
+  private static final class TMSLogPublisher extends AbstractExecutionThreadService {
+
+    private final BlockingQueue<LogMessage> messageQueue;
+    private final String topicPrefix;
+    private final int numPartitions;
+    private final int queueSize;
+    private final LoggingEventSerializer loggingEventSerializer;
+    private final ContextAware contextAware;
+    private final MessagingContext messagingContext;
+    private final LogPartitionType logPartitionType;
+    private final RetryStrategy retryStrategy;
+    private volatile Thread blockingThread;
+
+    private TMSLogPublisher(CConfiguration cConf, BlockingQueue<LogMessage> messageQueue, ContextAware contextAware,
+                            MessagingService messagingService, int queueSize) {
+      this.messageQueue = messageQueue;
+      this.topicPrefix = cConf.get(Constants.Logging.TMS_TOPIC_PREFIX);
+      this.numPartitions = cConf.getInt(Constants.Logging.NUM_PARTITIONS);
+      this.queueSize = queueSize;
+      this.loggingEventSerializer = new LoggingEventSerializer();
+      this.contextAware = contextAware;
+      this.retryStrategy = RetryStrategies.fromConfiguration(cConf, "system.log.process.");
+      this.logPartitionType =
+              LogPartitionType.valueOf(cConf.get(Constants.Logging.LOG_PUBLISH_PARTITION_KEY).toUpperCase());
+      this.messagingContext = new MultiThreadMessagingContext(messagingService);
+    }
+
+    @Override
+    protected void run() {
+      Map<Integer, List<byte[]>> buffer = new HashMap<>(queueSize);
+
+      int failures = 0;
+      long failureStartTime = System.currentTimeMillis();
+      while (isRunning()) {
```

Comment:
Extends from `AbstractRetryableScheduledService` instead to simplify the looping and retry logic.